server/datastore/mysql/mdm_test.go

```diff
@@ -3021,3 +3022,63 @@ func testBatchSetProfileLabelAssociations(t *testing.T, ds *Datastore) {
 		require.Error(t, err)
 	})
 }
+
+func testBatchSetMDMProfilesTransactionError(t *testing.T, ds *Datastore) {
+	ctx := context.Background()
+
+	lbl, err := ds.NewLabel(ctx, &fleet.Label{Name: "label", Query: "select 1"})
+	require.NoError(t, err)
+
+	cases := []struct {
+		windowsErr string
+		appleErr   string
+		wantErr    string
+	}{
+		{"select:a", "", "batch set windows profiles: load existing profiles: select:a"},
+		{"insert:b", "", ": insert:b"},
+		{"delete:c", "", "batch set windows profiles: delete obsolete profiles: delete:c"},
+		{"reselect:d", "", "batch set windows profiles: load newly inserted profiles: reselect:d"},
+		{"labels:e", "", "batch set windows profiles: inserting windows profile label associations: labels:e"},
+		{"", "select:f", "batch set apple profiles: load existing profiles: select:f"},
+		{"", "insert:g", ": insert:g"},
+		{"", "delete:h", "batch set apple profiles: delete obsolete profiles: delete:h"},
+		{"", "reselect:i", "batch set apple profiles: load newly inserted profiles: reselect:i"},
+		{"", "labels:j", "batch set apple profiles: inserting apple profile label associations: labels:j"},
+	}
```

Comment:
@roperzh despite injecting errors in every stages of the batch-set functions, I wasn't able to reproduce the "lost errors" you mention in the ticket. Regardless, it's a good thing to remove the unnecessary double-transactions as that would've hijacked two database connections for the duration of the processing, but if you have a sure way of reproducing those lost errors let me know and I'll add it to the test.

Reply:
@mna thanks and sorry if I wasted your time with that. In my case was an `sqlx.In` error around here:

https://github.com/fleetdm/fleet/blob/893afa6505169515005d7532e15662e5fbf6198e/server/datastore/mysql/microsoft_mdm.go#L1685-L1688

BUT, I was working on that ticket in a time crunch, so maybe the error wasn't bubbling for another reason, or I was just confused and I attributed it to the nested transactions.

Reply:
No worries! It's still worth cleaning up the double transactions. I'll add a test case for the "In" error, thanks!